<a href="https://colab.research.google.com/github/joshdsolis/DS-Unit-2-Sprint-3-Advanced-Regression/blob/master/DS_Unit_2_Sprint_Challenge_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 2 Sprint Challenge 3

## Logistic Regression and Beyond

In this sprint challenge you will fit a logistic regression modeling the probability of an adult having an income above 50K. The dataset is available at UCI:

https://archive.ics.uci.edu/ml/datasets/adult

Your goal is to:

1. Load, validate, and clean/prepare the data.
2. Fit a logistic regression model
3. Answer questions based on the results (as well as a few extra questions about the other modules)

Don't let the perfect be the enemy of the good! Manage your time, and make sure to get to all parts. If you get stuck wrestling with the data, simplify it (if necessary, drop features or rows) so you're able to move on. If you have time at the end, you can go back and try to fix/improve.

### Hints

It has a variety of features - some are continuous, but many are categorical. You may find [pandas.get_dummies](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html) (a method to one-hot encode) helpful!

The features have dramatically different ranges. You may find [sklearn.preprocessing.minmax_scale](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.minmax_scale.html#sklearn.preprocessing.minmax_scale) helpful!

## Part 1 - Load, validate, and prepare data

The data is available at: https://archive.ics.uci.edu/ml/datasets/adult

Load it, name the columns, and make sure that you've loaded the data successfully. Note that missing values for categorical variables can essentially be considered another category ("unknown"), and may not need to be dropped.

You should also prepare the data for logistic regression - one-hot encode categorical features as appropriate.

In [39]:
import pandas as pd

# TODO - your work!
names = ['ages', 'workclass', 'fnlwgt','education','education-num','marital-status','occupation','relationship','race','sex','capital-gain','capital-loss','hours-per-week','native-country','50K']
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data',header=None, names = names)
pd.set_option('display.max_columns', 100)
df.head()

ages          workclass  fnlwgt   education  education-num       marital-status          occupation    relationship    race      sex  capital-gain  capital-loss  hours-per-week  native-country     50K
0    39          State-gov   77516   Bachelors             13        Never-married        Adm-clerical   Not-in-family   White     Male          2174             0              40   United-States   <=50K
1    50   Self-emp-not-inc   83311   Bachelors             13   Married-civ-spouse     Exec-managerial         Husband   White     Male             0             0              13   United-States   <=50K
2    38            Private  215646     HS-grad              9             Divorced   Handlers-cleaners   Not-in-family   White     Male             0             0              40   United-States   <=50K
3    53            Private  234721        11th              7   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male             0             0              40   United-States   <=50K
4    28            Private  338409   Bachelors             13   Married-civ-spouse      Prof-specialty            Wife   Black   Female             0             0              40            Cuba   <=50K

In [40]:
df.columns

Index(['ages', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country', '50K'], dtype='object')

In [41]:
df['50K'].value_counts()

 <=50K    24720
 >50K      7841
Name: 50K, dtype: int64

In [0]:
rep = { ' <=50K': 0, ' >50K': 1}
df['50K'].replace(rep,inplace=True)


In [43]:
df.dtypes

ages               int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
50K                int64
dtype: object

In [44]:
df = pd.get_dummies(data=df,columns = ['workclass','education','occupation'])
df.head()

ages  fnlwgt  education-num       marital-status    relationship    race      sex  capital-gain  capital-loss  hours-per-week  native-country  50K  workclass_ ?  workclass_ Federal-gov  workclass_ Local-gov  workclass_ Never-worked  workclass_ Private  workclass_ Self-emp-inc  workclass_ Self-emp-not-inc  workclass_ State-gov  workclass_ Without-pay  education_ 10th  education_ 11th  education_ 12th  education_ 1st-4th  education_ 5th-6th  education_ 7th-8th  education_ 9th  education_ Assoc-acdm  education_ Assoc-voc  education_ Bachelors  education_ Doctorate  education_ HS-grad  education_ Masters  education_ Preschool  education_ Prof-school  education_ Some-college  occupation_ ?  occupation_ Adm-clerical  occupation_ Armed-Forces  occupation_ Craft-repair  occupation_ Exec-managerial  occupation_ Farming-fishing  occupation_ Handlers-cleaners  occupation_ Machine-op-inspct  occupation_ Other-service  occupation_ Priv-house-serv  occupation_ Prof-specialty  \
0    39   77516             13        Never-married   Not-in-family   White     Male          2174             0              40   United-States    0             0                       0                     0                        0                   0                        0                            0                     1                       0                0                0                0                   0                   0                   0               0                      0                     0                     1                     0                   0                   0                     0                       0                        0              0                         1                         0                         0                            0                            0                              0                              0                          0                            0                           0   
1    50   83311             13   Married-civ-spouse         Husband   White     Male             0             0              13   United-States    0             0                       0                     0                        0                   0                        0                            1                     0                       0                0                0                0                   0                   0                   0               0                      0                     0                     1                     0                   0                   0                     0                       0                        0              0                         0                         0                         0                            1                            0                              0                              0                          0                            0                           0   
2    38  215646              9             Divorced   Not-in-family   White     Male             0             0              40   United-States    0             0                       0                     0                        0                   1                        0                            0                     0                       0                0                0                0                   0                   0                   0               0                      0                     0                     0                     0                   1                   0                     0                       0                        0              0                         0                         0                         0                            0                            0                              1                              0                          0                            0                           0   
3    53  234721              7   Married-civ-spouse         Husband   Black

## Part 2 - Fit and present a Logistic Regression

Your data should now be in a state to fit a logistic regression. Use scikit-learn, define your `X` (independent variable) and `y`, and fit a model.

Then, present results - display coefficients in as interpretible a way as you can (hint - scaling the numeric features will help, as it will at least make coefficients more comparable to each other). If you find it helpful for interpretation, you can also generate predictions for cases (like our 5 year old rich kid on the Titanic) or make visualizations - but the goal is your exploration to be able to answer the question, not any particular plot (i.e. don't worry about polishing it).

It is *optional* to use `train_test_split` or validate your model more generally - that is not the core focus for this week. So, it is suggested you focus on fitting a model first, and if you have time at the end you can do further validation.

In [51]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import scale

# TODO - your work!
df_num = df._get_numeric_data()

X = df_num.drop(['50K'],axis=1)
X= scale(X)
X = pd.DataFrame(X)
y = df_num['50K']

absent_log1 = LogisticRegression().fit(X,y)
absent_log1.score(X,y)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:8: DataConversionWarning: Data with input dtype uint8, int64 were all converted to float64 by the scale function.
  
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


0.8227020054666626

In [52]:
absent_log1.coef_

array([[ 0.59316229,  0.05100097,  0.36150098,  2.34833543,  0.2686309 ,
         0.44025714, -0.07216602,  0.08709165, -0.03168189, -0.07025183,
         0.03741608,  0.09506864, -0.05096695, -0.03774935, -0.13120697,
        -0.09511349, -0.1284113 , -0.04813409, -0.05144633, -0.04715407,
        -0.0976483 , -0.0738455 , -0.00601123,  0.02875883,  0.13714418,
         0.09948229, -0.04468644,  0.11343868, -0.46390299,  0.12549587,
         0.01722096, -0.07649412, -0.11062105, -0.01752259,  0.13451474,
         0.25366741, -0.14253411, -0.11915547, -0.02056221, -0.35169035,
        -0.34745103,  0.12430033,  0.10926625,  0.08654037,  0.08967179,
         0.0411045 ]])

## Part 3 - Analysis, Interpretation, and Questions

### Based on your above model, answer the following questions

1. What are 3 features positively correlated with income above 50k?
2. What are 3 features negatively correlated with income above 50k?
3. Overall, how well does the model explain the data and what insights do you derive from it?

*These answers count* - that is, make sure to spend some time on them, connecting to your analysis above. There is no single right answer, but as long as you support your reasoning with evidence you are on the right track.

Note - scikit-learn logistic regression does *not* automatically perform a hypothesis test on coefficients. That is OK - if you scale the data they are more comparable in weight.

### Match the following situation descriptions with the model most appropriate to addressing them

In addition to logistic regression, a number of other approaches were covered this week. Pair them with the situations they are most appropriate for, and briefly explain why.

Situations:
1. You are given data on academic performance of primary school students, and asked to fit a model to help predict "at-risk" students who are likely to receive the bottom tier of grades.
2. You are studying tech companies and their patterns in releasing new products, and would like to be able to model and predict when a new product is likely to be launched.
3. You are working on modeling expected plant size and yield with a laboratory that is able to capture fantastically detailed physical data about plants, but only of a few dozen plants at a time.

Approaches:
1. Ridge Regression
2. Quantile Regression
3. Survival Analysis

In [48]:
df_num.shape

(32561, 47)

In [49]:
df_num.head()

ages  fnlwgt  education-num  capital-gain  capital-loss  hours-per-week  50K  workclass_ ?  workclass_ Federal-gov  workclass_ Local-gov  workclass_ Never-worked  workclass_ Private  workclass_ Self-emp-inc  workclass_ Self-emp-not-inc  workclass_ State-gov  workclass_ Without-pay  education_ 10th  education_ 11th  education_ 12th  education_ 1st-4th  education_ 5th-6th  education_ 7th-8th  education_ 9th  education_ Assoc-acdm  education_ Assoc-voc  education_ Bachelors  education_ Doctorate  education_ HS-grad  education_ Masters  education_ Preschool  education_ Prof-school  education_ Some-college  occupation_ ?  occupation_ Adm-clerical  occupation_ Armed-Forces  occupation_ Craft-repair  occupation_ Exec-managerial  occupation_ Farming-fishing  occupation_ Handlers-cleaners  occupation_ Machine-op-inspct  occupation_ Other-service  occupation_ Priv-house-serv  occupation_ Prof-specialty  occupation_ Protective-serv  occupation_ Sales  occupation_ Tech-support  \
0    39   77516             13          2174             0              40    0             0                       0                     0                        0                   0                        0                            0                     1                       0                0                0                0                   0                   0                   0               0                      0                     0                     1                     0                   0                   0                     0                       0                        0              0                         1                         0                         0                            0                            0                              0                              0                          0                            0                           0                            0                  0                         0   
1    50   83311             13             0             0              13    0             0                       0                     0                        0                   0                        0                            1                     0                       0                0                0                0                   0                   0                   0               0                      0                     0                     1                     0                   0                   0                     0                       0                        0              0                         0                         0                         0                            1                            0                              0                              0                          0                            0                           0                            0                  0                         0   
2    38  215646              9             0             0              40    0             0                       0                     0                        0                   1                        0                            0                     0                       0                0                0                0                   0                   0                   0               0                      0                     0                     0                     0                   1                   0                     0                       0                        0              0                         0                         0                         0                            0                            0                              1                              0                          0                            0                           0                            0                  0                         0   
3    53  234721              7             0             0 

**TODO - your answers!**

1) Age, capital gain, and education-num

2) Work class never worked, work class local-gov, and education 11th

3) The mean accuracy for the logistic regression is 0.823, the insights I get are whether a feature is positively or negatively correlated through the coefficients. There is a rather high impact for having capital gain as well as a higher than average impact for age.

1) Quantile regression, because we are seeking a specific quantile (those students receving bottom tier of grades). Using quantile regression would allow us to predict grade as a function of other features while targeting a cutoff.

2) Survival analysis, because what is wanted is the time-to-launch of a new product. Survival anlaysis is referred to as time-to-event analysis.

3) Ridge regression, because the small amount of plants at a time is built for ridge regression as it specializes in fewer observation models.